modules/unsupported/csv/src/main/java/org/geotools/data/csv/CSVFeatureWriter.java

```diff
@@ -0,0 +1,174 @@
+ *
+ * This file is hereby placed into the Public Domain. This means anyone is
+ * free to do whatever they wish with this file. Use it well and enjoy!
+ */
+package org.geotools.data.csv;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.NoSuchElementException;
+
+import org.geotools.data.DataUtilities;
+import org.geotools.data.FeatureWriter;
+import org.geotools.data.Query;
+import org.geotools.data.csv.parse.CSVIterator;
+import org.geotools.data.csv.parse.CSVStrategy;
+import org.geotools.feature.simple.SimpleFeatureBuilder;
+import org.opengis.feature.simple.SimpleFeature;
+import org.opengis.feature.simple.SimpleFeatureType;
+
+import com.csvreader.CsvWriter;
+
+/**
+ * Iterator supporting writing of feature content.
+ *
+ * @author Jody Garnett (Boundless)
+ * @author Lee Breisacher
+ */
+public class CSVFeatureWriter implements FeatureWriter<SimpleFeatureType, SimpleFeature> {
+    private SimpleFeatureType featureType;
+    private CSVStrategy csvStrategy;
+    private CSVFileState csvFileState;
+    
+    /** Temporary file used to stage output */
+    private File temp;
+    
+    /** iterator handling reading of original file(?) */
+    private CSVIterator iterator;
+    
+    /** CsvWriter used for temp file output */
+    private CsvWriter csvWriter;
+    
+    /** Flag indicating we have reached the end of the file */
+    private boolean appending = false;
+    
+    /** Row count used to generate FeatureId when appending */
+    int nextRow = 0;
+    
+    /** Current feature available for modification. May be null if feature removed */
+    private SimpleFeature currentFeature;
+
+    public CSVFeatureWriter(CSVFileState csvFileState, CSVStrategy csvStrategy)
+    		throws IOException {
+        this(csvFileState, csvStrategy, Query.ALL);
+    }
+    
+    public CSVFeatureWriter(CSVFileState csvFileState, CSVStrategy csvStrategy, Query query)
+    		throws IOException {
+    	this.csvFileState = csvFileState;
+    	File file = csvFileState.getFile();
+    	File directory = file.getParentFile();
+    	String typeName = query.getTypeName();
+    	this.temp = File.createTempFile(typeName + System.currentTimeMillis(), "csv", directory);
+    	this.featureType = csvStrategy.getFeatureType();
+    	this.iterator = csvStrategy.iterator();
+    	this.csvStrategy = csvStrategy;
+    	this.csvWriter = new CsvWriter(new FileWriter(this.temp), ',');
+    	this.csvWriter.writeRecord(this.csvFileState.getCSVHeaders());
+    }
+    
+    // featureType start
+    @Override
+    public SimpleFeatureType getFeatureType() {
+        return this.featureType;
+    }
+    // featureType end
+
+    // hasNext start
+    @Override
+    public boolean hasNext() throws IOException {
+        if( csvWriter == null ){
+            throw new IOException("Writer has been closed");
```

Comment:
@tbarsballe has gone through and done a QA run to ensure closed writers return hasNext(): false (rather than thrown an IOException).

Lets stick with that for consistency (principle of least surprise)